src/NLog/Internal/CallSiteInformation.cs

```diff
@@ -0,0 +1,146 @@
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+namespace NLog.Internal
+{
+    using System;
+    using System.Diagnostics;
+    using System.Reflection;
+
+    internal class CallSiteInformation
+    {
+        /// <summary>
+        /// Sets the stack trace for the event info.
+        /// </summary>
+        /// <param name="stackTrace">The stack trace.</param>
+        /// <param name="userStackFrame">Index of the first user stack frame within the stack trace.</param>
+        /// <param name="userStackFrameLegacy">Index of the first user stack frame within the stack trace.</param>
+        public void SetStackTrace(StackTrace stackTrace, int userStackFrame, int? userStackFrameLegacy)
+        {
+            StackTrace = stackTrace;
+            UserStackFrameNumber = userStackFrame;
+            UserStackFrameNumberLegacy = userStackFrameLegacy != userStackFrame ? userStackFrameLegacy : null;
+        }
+
+        /// <summary>
+        /// Sets the details retrieved from the Caller Information Attributes
+        /// </summary>
+        /// <param name="callerMemberName"></param>
+        /// <param name="callerFilePath"></param>
+        /// <param name="callerLineNumber"></param>
+        public void SetCallerInfo(string callerMemberName, string callerFilePath, int callerLineNumber)
+        {
+            CallerMemberName = callerMemberName;
+            CallerFilePath = callerFilePath;
+            CallerLineNumber = callerLineNumber;
+        }
+
+        /// <summary>
+        /// Gets the stack frame of the method that did the logging.
+        /// </summary>
+        public StackFrame UserStackFrame => (StackTrace != null) ? StackTrace.GetFrame(UserStackFrameNumberLegacy ?? UserStackFrameNumber) : null;
+
+        /// <summary>
+        /// Gets the number index of the stack frame that represents the user
+        /// code (not the NLog code).
+        /// </summary>
+        public int UserStackFrameNumber { get; private set; }
+
+        /// <summary>
+        /// Legacy attempt to skip async MoveNext, but caused source file line number to be lost
+        /// </summary>
+        public int? UserStackFrameNumberLegacy { get; private set; }
+
+        /// <summary>
+        /// Gets the entire stack trace.
+        /// </summary>
+        public StackTrace StackTrace { get; private set; }
+
+        public MethodBase GetCallerStackFrameMethod(int skipFrames)
+        {
+            StackFrame frame = StackTrace?.GetFrame(UserStackFrameNumber + skipFrames);
+            return frame?.GetMethod();
+        }
+
+        public string GetCallerClassName(MethodBase method, bool includeNameSpace, bool cleanAsyncMoveNext, bool cleanAnonymousDelegates)
```

Comment:
What do you think about a struct/class for the last 3 parameters?

(the same object for `GetCallerMemberName`

Only for readability of the calling site.

Reply:
Have no opinion. But not that big fan of struct-parameters.

Reply:
OK,it's internal.so could be done in the future